Unable to use surefire maven plugin to rerun failing tests

## Summary

Using [surefire to rerun failing tests ](https://maven.apache.org/surefire/maven-surefire-plugin/examples/rerun-failing-tests.html) results in a stack trace.

- cucumber-java8:1.2.5
- maven-failsafe-plugin:2.20
- maven-surefire-plugin:2.20 


## Expected Behavior

When using surefire with `mvn -Dfailsafe.rerunFailingTestsCount=2 test` I expect the failing tests to be rerun.

## Current Behavior

```
org.apache.maven.surefire.testset.TestSetFailedException: Unable to create test class 'Scenario: Fail when running'
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeFailedMethod(JUnit4Provider.java:385)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:292)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
Caused by: java.lang.ClassNotFoundException: Scenario: Fail when running
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeFailedMethod(JUnit4Provider.java:379)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:292)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
```

## Possible Solution

Surefire parses the test description for the class to rerun and expects a class name here rather then a human readable name. Either provide a class name or help convince the surefire folks to fix this this. :)

See: [SUREFIRE-1372](https://issues.apache.org/jira/browse/SUREFIRE-1372)

## Steps to Reproduce (for bugs)

1. Setup mvn project with a failing cucumber test
2. Add a test runner that uses `@RunWith(Cucumber.class)` and `@CucumberOptions(...)`
3. Run mvn -Dfailsafe.rerunFailingTestsCount=2 test
4. Observe stacktrace.

## Context & Motivation

Rerunning failing tests as an integral part of the build, especially in large tests  sets that are ran overnight, allows us to prioritize investigating the genuine failures over the flaky ones.

--- a/junit/src/main/java/cucumber/runtime/junit/ExecutionUnitRunner.java
+++ b/junit/src/main/java/cucumber/runtime/junit/ExecutionUnitRunner.java
@@ -131,10 +131,10 @@ class ExecutionUnitRunner extends ParentRunner<PickleStep> {
         private final int pickleLine;
         private final int pickleStepLine;
 
-        PickleStepId(PickleEvent pickleEvent, PickleStep pickleStepLine) {
+        PickleStepId(PickleEvent pickleEvent, PickleStep pickleStep) {
             this.uri = pickleEvent.uri;
             this.pickleLine = pickleEvent.pickle.getLocations().get(0).getLine();
-            this.pickleStepLine = pickleStepLine.getLocations().get(0).getLine();
+            this.pickleStepLine = pickleStep.getLocations().get(0).getLine();
         }
 
         @Override

--- a/junit/src/main/java/cucumber/runtime/junit/ExecutionUnitRunner.java
+++ b/junit/src/main/java/cucumber/runtime/junit/ExecutionUnitRunner.java
@@ -49,7 +49,7 @@ class ExecutionUnitRunner extends ParentRunner<PickleStep> {
     public Description getDescription() {
         if (description == null) {
             String nameForDescription = getName().isEmpty() ? "EMPTY_NAME" : getName();
-            description = Description.createSuiteDescription(nameForDescription, new PickleWrapper(pickleEvent));
+            description = Description.createSuiteDescription(nameForDescription, new PickleId(pickleEvent));
 
             for (PickleStep step : getChildren()) {
                 description.addChild(describeChild(step));
@@ -68,7 +68,7 @@ class ExecutionUnitRunner extends ParentRunner<PickleStep> {
             } else {
                 testName = step.getText();
             }
-            description = Description.createTestDescription(getName(), testName, new PickleStepWrapper(step));
+            description = Description.createTestDescription(getName(), testName, new PickleStepId(pickleEvent, step));
             stepDescriptions.put(step, description);
         }
         return description;
@@ -93,22 +93,69 @@ class ExecutionUnitRunner extends ParentRunner<PickleStep> {
     private String makeNameFilenameCompatible(String name) {
         return name.replaceAll("[^A-Za-z0-9_]", "_");
     }
-}
 
-class PickleWrapper implements Serializable {
-    private static final long serialVersionUID = 1L;
-    private PickleEvent pickleEvent;
+    private static final class PickleId implements Serializable {
+        private static final long serialVersionUID = 1L;
+        private final String uri;
+        private int pickleLine;
 
-    PickleWrapper(PickleEvent pickleEvent) {
-        this.pickleEvent = pickleEvent;
+        PickleId(PickleEvent pickleEvent) {
+            this.uri = pickleEvent.uri;
+            this.pickleLine = pickleEvent.pickle.getLocations().get(0).getLine();
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            PickleId that = (PickleId) o;
+            return pickleLine == that.pickleLine && uri.equals(that.uri);
+        }
+
+        @Override
+        public int hashCode() {
+            int result = uri.hashCode();
+            result = 31 * result + pickleLine;
+            return result;
+        }
+
+        @Override
+        public String toString() {
+            return uri + ":" + pickleLine;
+        }
     }
-}
 
-class PickleStepWrapper implements Serializable {
-    private static final long serialVersionUID = 1L;
-    private PickleStep step;
+    private static final class PickleStepId implements Serializable {
+        private static final long serialVersionUID = 1L;
+        private final String uri;
+        private final int pickleLine;
+        private final int pickleStepLine;
 
-    PickleStepWrapper(PickleStep step) {
-        this.step = step;
+        PickleStepId(PickleEvent pickleEvent, PickleStep pickleStepLine) {
+            this.uri = pickleEvent.uri;
+            this.pickleLine = pickleEvent.pickle.getLocations().get(0).getLine();
+            this.pickleStepLine = pickleStepLine.getLocations().get(0).getLine();
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            PickleStepId that = (PickleStepId) o;
+            return pickleLine == that.pickleLine && pickleStepLine == that.pickleStepLine && uri.equals(that.uri);
+        }
+
+        @Override
+        public int hashCode() {
+            int result = pickleLine;
+            result = 31 * result + uri.hashCode();
+            result = 31 * result + pickleStepLine;
+            return result;
+        }
+
+        @Override
+        public String toString() {
+            return uri + ":" + pickleLine + ":" + pickleStepLine;
+        }
     }
 }

--- a/junit/src/main/java/cucumber/runtime/junit/FeatureRunner.java
+++ b/junit/src/main/java/cucumber/runtime/junit/FeatureRunner.java
@@ -12,6 +12,7 @@ import org.junit.runner.notification.RunNotifier;
 import org.junit.runners.ParentRunner;
 import org.junit.runners.model.InitializationError;
 
+import java.io.Serializable;
 import java.util.ArrayList;
 import java.util.List;
 
@@ -36,7 +37,7 @@ public class FeatureRunner extends ParentRunner<ParentRunner> {
     @Override
     public Description getDescription() {
         if (description == null) {
-            description = Description.createSuiteDescription(getName(), cucumberFeature);
+            description = Description.createSuiteDescription(getName(), new FeatureId(cucumberFeature));
             for (ParentRunner child : getChildren()) {
                 description.addChild(describeChild(child));
             }
@@ -84,6 +85,34 @@ public class FeatureRunner extends ParentRunner<ParentRunner> {
                     throw new CucumberException("Failed to create scenario runner", e);
                 }
             }
+        }
+    }
+
+    private static final class FeatureId implements Serializable {
+        private static final long serialVersionUID = 1L;
+        private final String uri;
+
+        FeatureId(CucumberFeature feature) {
+            this.uri = feature.getPath();
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            FeatureId featureId = (FeatureId) o;
+            return uri.equals(featureId.uri);
+
+        }
+
+        @Override
+        public int hashCode() {
+            return uri.hashCode();
+        }
+
+        @Override
+        public String toString() {
+            return uri;
         }
     }
 

--- a/junit/src/test/java/cucumber/runtime/junit/FeatureRunnerTest.java
+++ b/junit/src/test/java/cucumber/runtime/junit/FeatureRunnerTest.java
@@ -7,6 +7,7 @@ import cucumber.runtime.RuntimeGlue;
 import cucumber.runtime.RuntimeOptions;
 import cucumber.runtime.io.ClasspathResourceLoader;
 import cucumber.runtime.model.CucumberFeature;
+import org.junit.Assert;
 import org.junit.Test;
 import org.junit.runner.Description;
 import org.junit.runner.notification.RunNotifier;
@@ -15,8 +16,11 @@ import org.mockito.ArgumentMatcher;
 import org.mockito.InOrder;
 
 import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
 
 import static java.util.Arrays.asList;
+import static org.junit.Assert.assertTrue;
 import static org.mockito.Matchers.argThat;
 import static org.mockito.Mockito.inOrder;
 import static org.mockito.Mockito.mock;
@@ -89,17 +93,67 @@ public class FeatureRunnerTest {
     }
 
     private RunNotifier runFeatureWithNotifier(CucumberFeature cucumberFeature) throws InitializationError {
-        final RuntimeOptions runtimeOptions = new RuntimeOptions("-p null");
-        final ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
-        final ClasspathResourceLoader resourceLoader = new ClasspathResourceLoader(classLoader);
-        final RuntimeGlue glue = mock(RuntimeGlue.class);
-        final Runtime runtime = new Runtime(resourceLoader, classLoader, asList(mock(Backend.class)), runtimeOptions, new TimeService.Stub(0l), glue);
-        FeatureRunner runner = new FeatureRunner(cucumberFeature, runtime, new JUnitReporter(runtime.getEventBus(), false, new JUnitOptions(Collections.<String>emptyList())));
+        FeatureRunner runner = createFeatureRunner(cucumberFeature);
         RunNotifier notifier = mock(RunNotifier.class);
         runner.run(notifier);
         return notifier;
     }
 
+    private FeatureRunner createFeatureRunner(CucumberFeature cucumberFeature) throws InitializationError {
+        final RuntimeOptions runtimeOptions = new RuntimeOptions("-p null");
+        final ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+        final ClasspathResourceLoader resourceLoader = new ClasspathResourceLoader(classLoader);
+        final RuntimeGlue glue = mock(RuntimeGlue.class);
+        final Runtime runtime = new Runtime(resourceLoader, classLoader, asList(mock(Backend.class)), runtimeOptions, new TimeService.Stub(0l), glue);
+        return new FeatureRunner(cucumberFeature, runtime, new JUnitReporter(runtime.getEventBus(), false, new JUnitOptions(Collections.<String>emptyList())));
+    }
+
+
+    @Test
+    public void shouldPopulateDescriptionsWithStableUniqueIds() throws Exception {
+        CucumberFeature cucumberFeature = TestPickleBuilder.parseFeature("path/test.feature", "" +
+            "Feature: feature name\n" +
+            "  Background:\n" +
+            "    Given background step\n" +
+            "  Scenario: A\n" +
+            "    Then scenario name\n" +
+            "  Scenario: B\n" +
+            "    Then scenario name\n" +
+            "  Scenario Outline: C\n" +
+            "    Then scenario <name>\n" +
+            "  Examples:\n" +
+            "    | name |\n" +
+            "    | C    |\n" +
+            "    | D    |\n" +
+            "    | E    |\n"
+
+        );
+
+        FeatureRunner runner = createFeatureRunner(cucumberFeature);
+        FeatureRunner rerunner = createFeatureRunner(cucumberFeature);
+
+        Set<Description> descriptions = new HashSet<Description>();
+        assertDescriptionIsUnique(runner.getDescription(), descriptions);
+        assertDescriptionIsPredictable(runner.getDescription(), descriptions);
+        assertDescriptionIsPredictable(rerunner.getDescription(), descriptions);
+
+    }
+
+    private static void assertDescriptionIsUnique(Description description, Set<Description> descriptions) {
+        // Note, JUnit uses the the serializable parameter (in this case the step)
+        // as the unique id when comparing Descriptions
+        assertTrue(descriptions.add(description));
+        for (Description each : description.getChildren()) {
+            assertDescriptionIsUnique(each, descriptions);
+        }
+    }
+
+    private static void assertDescriptionIsPredictable(Description description, Set<Description> descriptions) {
+        assertTrue(descriptions.contains(description));
+        for (Description each : description.getChildren()) {
+            assertDescriptionIsPredictable(each, descriptions);
+        }
+    }
 }
 
 class DescriptionMatcher extends ArgumentMatcher<Description> {

